tools/raid_xmltolua.cpp

```diff
@@ -0,0 +1,479 @@
+#include <getopt.h>
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <unordered_map>
+#include <vector>
+
+#include <boost/filesystem.hpp>
+#include <pugixml.hpp>
+#include "../src/pugicast.h"
+
+namespace fs = boost::filesystem;
+
+void usage(const char* program)
+{
+	std::cout << "usage: " << program
+		<< " [-r path/to/data/raids] [-g path/to/data/globalevents]\n";
+}
+
+constexpr const char* INDENT = "    ";
+
+struct RaidEvent
+{
+	explicit RaidEvent(uint32_t delay) : delay{delay} {}
+	virtual ~RaidEvent() = default;
+
+	virtual void to_script(std::ostringstream& ss) const { ss << '\n'; };
+
+	uint32_t delay;
+};
+
+static const std::unordered_map<std::string, std::string> messageTypes{
+	{"warning", "MESSAGE_STATUS_WARNING"},
+	{"event", "MESSAGE_EVENT_ADVANCE"},
+	{"default", "MESSAGE_EVENT_DEFAULT"},
+	{"description", "MESSAGE_INFO_DESCR"},
+	{"smallstatus", "MESSAGE_STATUS_SMALL"},
+	{"blueconsole", "MESSAGE_STATUS_CONSOLE_BLUE"},
+	{"redconsole", "MESSAGE_STATUS_CONSOLE_RED"},
+};
+static const std::string defaultMessageType{"MESSAGE_EVENT_ADVANCE"};
+
+struct AnnounceEvent : public RaidEvent
+{
+	AnnounceEvent(uint32_t delay, std::string msg, std::string type) :
+			RaidEvent{delay}, message{std::move(msg)}, type{std::move(type)} {}
+
+	static AnnounceEvent* from(uint32_t delay, const pugi::xml_node& eventNode)
+	{
+		pugi::xml_attribute message = eventNode.attribute("message");
+		if (not message) {
+			throw std::runtime_error("message tag missing for announce event.");
+		}
+
+		std::string messageType = [](const pugi::xml_attribute& type) {
+			if (not type) {
+				std::cout << "[Notice] Raid: type tag missing for announce"
+				             "event. Using default: " << defaultMessageType <<
+							 "\n";
+				return defaultMessageType;
+			}
+
+			try {
+				return messageTypes.at(type.as_string());
+			} catch (const std::out_of_range&) {
+				std::cout << "[Notice] Raid: Unknown type tag for announce"
+							 "event. Using default: MESSAGE_TYPE_ADVANCE\n";
+				return defaultMessageType;
+			}
+		}(eventNode.attribute("type"));
+
+		return new AnnounceEvent{delay, message.as_string(), messageType};
+	}
+
+	void to_script(std::ostringstream& ss) const override
+	{
+		ss << INDENT << "Game.broadcastMessage(\"" << message << "\", " << type
+		   << ")\n";
+	}
+
+	std::string message, type;
+};
+
+struct Position
+{
+	int32_t x;
+	int32_t y;
+	int16_t z;
+};
+
+template<class T>
+T require_attr(const pugi::xml_node& node, const std::string& tag)
+{
+	if (pugi::xml_attribute attr = node.attribute(tag.c_str())) {
+		return pugi::cast<T>(attr.value());
+	}
+	throw std::runtime_error(tag + " tag missing.");
```

Comment:
Minor issue: I'd introduce a separate type of exception (`MissingTagError`) that inherits from a `std::runtime_error` to make it clear at catch point that other `std::runtime_errors` are unexpected and fatal.